tests/e2e/stateapp/stateapp_test.go

```diff
@@ -0,0 +1,297 @@
+type requestResponse struct {
+	States []daprState `json:"states,omitempty"`
+}
+
+// represents each step in a test since it can make multiple calls.
+type testStep struct {
+	command          string
+	request          requestResponse
+	expectedResponse requestResponse
+}
+
+type testCase struct {
+	name  string
+	steps []testStep
+}
+
+// internal key value structure to simplify code given bloated app API.
+type simpleKeyValue struct {
+	key   string
+	value *string
+}
+
+// creates a requestResponse based on an array of key value pairs.
+func newRequestResponse(keyValues ...simpleKeyValue) requestResponse {
+	var daprStates = make([]daprState, 0, len(keyValues))
+	for _, keyValue := range keyValues {
+		if keyValue.value != nil {
+			daprStates = append(daprStates, daprState{
+				keyValue.key,
+				&appState{
+					*keyValue.value,
+				},
+			})
+		} else {
+			daprStates = append(daprStates, daprState{
+				keyValue.key,
+				nil,
+			})
+		}
+	}
+	return requestResponse{
+		daprStates,
+	}
+}
+
+// Just for readability.
+func newRequest(keyValues ...simpleKeyValue) requestResponse {
+	return newRequestResponse(keyValues...)
+}
+
+// Just for readability.
+func newResponse(keyValues ...simpleKeyValue) requestResponse {
+	return newRequestResponse(keyValues...)
+}
+
+// Generates random keys (not values).
+func generateRandomKeysOnly(num int) []simpleKeyValue {
+	if (num < 0) || (num > maxEntries) {
+		return make([]simpleKeyValue, 0)
+	}
+
+	var output = make([]simpleKeyValue, 0, num)
+	for i := 1; i <= num; i++ {
+		var key = guuid.New().String()
+		output = append(output, simpleKeyValue{key, nil})
+	}
+
+	return output
+}
+
+// Generates values for the keys passed in.
+func generateRandomValues(keyValues []simpleKeyValue) []simpleKeyValue {
+	var output = make([]simpleKeyValue, 0, len(keyValues))
+	for i, keyValue := range keyValues {
+		var key = keyValue.key
+		var value = fmt.Sprintf("Value for entry #%d with key %s.", i, key)
+		output = append(output, simpleKeyValue{key, &value})
+	}
+
+	return output
+}
+
+func generateRandomKeys(num int) []simpleKeyValue {
+	if (num < 0) || (num > maxEntries) {
+		return make([]simpleKeyValue, 0)
+	}
+
+	var output = make([]simpleKeyValue, 0, num)
+	for i := 1; i <= num; i++ {
+		var key = guuid.New().String()
+		output = append(output, simpleKeyValue{key, nil})
+	}
+
+	return output
+}
+
+func generateTestCases() []testCase {
+	// Just for readability
+	var emptyRequest = requestResponse{
+		nil,
+	}
+
+	// Just for readability
+	var emptyResponse = requestResponse{
+		nil,
+	}
+
+	var testCase1Key = guuid.New().String()
+	var testCase1Value = "The best song ever is 'Highwayman' by 'The Highwaymen'."
+
+	var testCaseManyKeys = generateRandomKeys(testManyEntriesCount)
+	var testCaseManyKeyValues = generateRandomValues(testCaseManyKeys)
+
+	return []testCase{
+		{
+			// No comma since this will become the name of the test without spaces.
+			"Test get save delete with empty request response",
```

Comment:
It may help readability to indicate there is only a single app, no multi-hop.

Reply:
Done.